### ISSUE TITLE
com.hierynomus.mssmb2.SMBApiException: STATUS_OTHER (0xffffffff): Cannot resolve path

### ISSUE THREAD

**Issue by cliviu** [hierynomus/smbj#419]
Hi, I have the following exception:
2018-12-14 16:38:04,408|DFSPathResolver               |INFO |Starting DFS resolution for \\host1.com\Share\Resource
2018-12-14 16:38:04,408|ProxySocketFactory            |DEBUG|Connecting to \host1.com\Share:445
2018-12-14 16:38:04,424|Connection                    |INFO |Closed connection to \host1.com\Share
2018-12-14 16:38:04,424|SMBClient                     |DEBUG|Connection to << \host1.com\Share:445 >> closed
       |ERROR|FileCopy:STATUS_OTHER (0xffffffff): Cannot resolve path \\host1.com\Share\Resource
com.hierynomus.mssmb2.SMBApiException: STATUS_OTHER (0xffffffff): Cannot resolve path \\host1.com\Share\Resource
	at com.hierynomus.smbj.share.DiskShare.resolveAndCreateFile(DiskShare.java:81)
	at com.hierynomus.smbj.share.DiskShare.open(DiskShare.java:66)
	at com.hierynomus.smbj.share.DiskShare.exists(DiskShare.java:187)
	at com.hierynomus.smbj.share.DiskShare.folderExists(DiskShare.java:183)
	...
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hierynomus.msdfsc.DFSException: java.net.UnknownHostException: \host1.com\Share
	at com.hierynomus.smbj.paths.DFSPathResolver.sendDfsReferralRequest(DFSPathResolver.java:397)
	at com.hierynomus.smbj.paths.DFSPathResolver.step9(DFSPathResolver.java:306)
	at com.hierynomus.smbj.paths.DFSPathResolver.step2(DFSPathResolver.java:140)
	at com.hierynomus.smbj.paths.DFSPathResolver.step1(DFSPathResolver.java:120)
	at com.hierynomus.smbj.paths.DFSPathResolver.resolve(DFSPathResolver.java:107)
	at com.hierynomus.smbj.paths.DFSPathResolver.resolve(DFSPathResolver.java:95)
	at com.hierynomus.smbj.share.DiskShare.resolveAndCreateFile(DiskShare.java:77)
	... 9 more

in DfsPathResolver, Line 395, there is a TODO:    
`connection = oldConnection.getClient().connect(hostName); // TODO`

Does it ring any bells  ?  

The strange thing is that  two minutes before, the same resolution was successful.
2018-12-14 16:36:04,533|DFSPathResolver               |INFO |Starting DFS resolution for \\host1.com\Share\Resource
2018-12-14 16:36:04,533|DFSPathResolver               |INFO |DFS resolved\\host1.com\Share\Resource -> \\host2.com\Share$\Test\...

Thanks,
   Liviu

**Comment by hierynomus** [comment]
The cached referral has expired in your case (see: DFSPathResolver#140). Will dig into that.

**Comment by cliviu** [comment]
Thanks !!

**Comment by Disablez** [comment]
Same problem here.

**Comment by vjorge** [comment]
Hi, facing the same issue.
Anyone can share a workaround?
I have an ISO file that needs to be kept open for random read, and after +-15 minutes it throws the exact same error above.
Using the latest code. 
Many thanks,

**Comment by vjorge** [comment]
Hi I found what is causing the issue:
The function sendDfsReferralRequest, is expecting the hostName as a string with the server name, but instead the functions 
stepX and passing the hostname with the format " \\host1.com\Share"

Then in function sendDfsReferralRequest it triggers the condition:

` if (!hostName.equals(session.getConnection().getRemoteHostname())) {`

Because the provided hostName is a UNC path not the host FQDN thus causing the error:

Caused by: com.hierynomus.msdfsc.DFSException: java.net.UnknownHostException: \host1.com\Share


I have not found what is causing the hostName to be passed with the format " \\host1.com\Share"

@[hierynomus](https://github.com/hierynomus) can you please help with this?

```
private ReferralResult sendDfsReferralRequest(DfsRequestType type, String hostName, Session session, DFSPath path) throws DFSException {
        // The client MUST initiate a server session with the SMB server, as specified in [MS-CIFS] section 3.4.4.7,
        // by passing HostName and UserCredentials as input parameters and receiving an opaque ClientGenericContext,
        // as specified in [MS-CIFS] section 3.4.
        // The client MUST search for an existing Session and TreeConnect to any share on the server identified by
        // ServerName for the user identified by UserCredentials. If no Session and TreeConnect are found, the client
        // MUST establish a new Session and TreeConnect to IPC$ on the target server as described in section 3.2.4.2
        // using the supplied ServerName and UserCredentials
        Session dfsSession = session;
        if (!hostName.equals(session.getConnection().getRemoteHostname())) {
            AuthenticationContext auth = session.getAuthenticationContext();
            Connection oldConnection = session.getConnection();
            Connection connection;
            try {
                connection = oldConnection.getClient().connect(hostName); // TODO
            } catch (IOException e) {
                throw new DFSException(e);
            }
            dfsSession = connection.authenticate(auth);
        }

        try {
            Share dfsShare = dfsSession.connectShare("IPC$"); // explicitly not closed as we want to re-use the cached Share for multiple requests
            return getReferral(type, dfsShare, path);
        } catch (Buffer.BufferException | IOException e) {
            throw new DFSException(e);
        }
    }
```

**Comment by vjorge** [comment]
Seems that found the issue:

The problem is with function step9 in :

`ReferralResult result = sendDfsReferralRequest(DfsRequestType.LINK, rootReferralCacheEntry.getTargetHint().getTargetPath(), session, state.path);`


The call ` rootReferralCacheEntry.getTargetHint().getTargetPath()` return a UNCPath not a FQDN hostname, but the sendDfsReferralRequest expects a FQDN Hostname, thus causing the connection to fail.

Bellow is the changed function that works, but I believe the correct way to fix it is to have a field named hostName in ReferralCache that returns the server hostname, and use it instead of the getTargetPath().

   ```
private <T> T step9(Session session, ResolveState<T> state, ReferralCache.ReferralCacheEntry lookup) throws DFSException {
        logger.trace("DFS[9]: {}", state);
        DFSPath rootPath = new DFSPath(state.path.getPathComponents().subList(0, 2));
        ReferralCache.ReferralCacheEntry rootReferralCacheEntry = referralCache.lookup(rootPath);
        if (rootReferralCacheEntry == null) {
            logger.debug("Could not find referral cache entry for {}", rootPath);
            referralCache.clear(state.path);
            return step1(session, state);
        }
        
        /*
         * Alterado por mim, porque existe um bug ou falha no codigo que passa o unc em vez do hostname
         */
        String target = rootReferralCacheEntry.getTargetHint().getTargetPath();
        if (target.startsWith("\\")) {
        	target = SmbPath.parse(target).getHostname();
        }
        
       // ReferralResult result = sendDfsReferralRequest(DfsRequestType.LINK, rootReferralCacheEntry.getTargetHint().getTargetPath(), session, state.path);
        ReferralResult result = sendDfsReferralRequest(DfsRequestType.LINK, target, session, state.path);
        if (!NtStatus.isSuccess(result.status)) {
            return step14(session, state, result);
        }

        if (result.referralCacheEntry.isRoot()) {
            return step3(session, state, result.referralCacheEntry);
        }

        return step4(session, state, result.referralCacheEntry);
    }
```

**Comment by hierynomus** [comment]
Can you test with the fix in #722? I've changed the implementation slightly.

**Comment by vjorge** [comment]
Hi hierynomus,
Many thanks for your update.
Seems that now its fixed ;) It works great.
Thank very much.
Vitor

**Comment by hierynomus** [comment]
Merged #722
